File: src/main/java/com/otus/service/AnswerServiceImpl.java
package com.otus.service;

import com.otus.domain.QuizElement;
import com.otus.repository.CsvRepository;
import com.otus.utils.CollectionUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

import static com.otus.utils.PrintUtils.print;

@RequiredArgsConstructor
public class AnswerServiceImpl implements AnswerService {
    private final UserService userService;
    private final CsvRepository csvRepository;

    private Scanner scanner = new Scanner(System.in);

    private static List<Integer> userAnswers = new ArrayList<>();

    @Override
    public void showResult() {
        userService.showFullName();
        print("You result: " + getResult() + " of " + questionCount());
    }

    @Override
    public Integer askAnswer(int quizElementIndex) {
        int answer = scanner.nextInt();
        userAnswers.add(answer);

        return answer;
    }

    private Integer questionCount(){
        return getCorrectAnswers(csvRepository.getAllElements()).size();
    }

    private Long getResult() {
        List<Integer> correctAnswers = getCorrectAnswers(csvRepository.getAllElements());

        return CollectionUtils.equalElementsCountConsideringIndex(correctAnswers, userAnswers);
    }


    private List<Integer> getCorrectAnswers(List<QuizElement> allElements) {
        return allElements.stream()
                .map(QuizElement::getCorrectAnswer)
                .map(Integer::parseInt)
                .collect(Collectors.toList());
    }
}

File: src/main/java/com/otus/service/AnswerServiceImpl.kt
package com.otus.service

import com.otus.repository.CorrectAnswerRepository
import org.springframework.stereotype.Service

@Service
open class AnswerServiceImpl(
    private val userService: UserService,
    private val scanner: ConsoleInputService,
    private val printService: PrintService,
    private val correctAnswerRepo: CorrectAnswerRepository,
    private val resultCalculationService: ResultCalculationService

) : AnswerService {

    private val questionLabelToUserAnswer = mutableMapOf<String, Int>()

    override fun showResult() {
        userService.showFullName()
        printService.println("Your result is ${userCorrectAnswersAmount()} of ${questionsAmount()}")
    }

    private fun userCorrectAnswersAmount() =
        resultCalculationService.getUserCorrectAnswersAmount(questionLabelToUserAnswer)

    private fun questionsAmount() = correctAnswerRepo.getQuestionsAmount()

    override fun askAnswer(questionLabel: String): Int {
        val answer = scanner.nextInt()
        questionLabelToUserAnswer[questionLabel] = answer

        return answer
    }
}